src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/TriviaDataFactory.TriviaRewriter.vb

```diff
@@ -56,21 +56,29 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Formatting
                 Next pair
             End Sub
 
-            Private Function GetTrailingAndLeadingTrivia(pair As KeyValuePair(Of ValueTuple(Of SyntaxToken, SyntaxToken), TriviaData)) As ValueTuple(Of SyntaxTriviaList, SyntaxTriviaList)
+            Private Function GetTrailingAndLeadingTrivia(pair As KeyValuePair(Of ValueTuple(Of SyntaxToken, SyntaxToken), TriviaData)) As (SyntaxTriviaList, SyntaxTriviaList)
                 If pair.Key.Item1.Kind = 0 OrElse _lastToken = pair.Key.Item2 Then
-                    Return ValueTuple.Create(Of SyntaxTriviaList, SyntaxTriviaList)(SyntaxTriviaList.Empty,
-                                                                                    GetSyntaxTriviaList(GetTextSpan(pair.Key), pair.Value, _cancellationToken))
+                    Return (SyntaxTriviaList.Empty,
+                            GetSyntaxTriviaList(GetTextSpan(pair.Key), pair.Value, _cancellationToken))
                 End If
 
                 Dim vbTriviaData = TryCast(pair.Value, TriviaDataWithList)
                 If vbTriviaData IsNot Nothing Then
                     Dim triviaList = vbTriviaData.GetTriviaList(_cancellationToken)
                     Dim index = GetIndexForEndOfLeadingTrivia(triviaList)
 
-                    Return ValueTuple.Create(SyntaxFactory.TriviaList(CreateTriviaListFromTo(triviaList, 0, index)), SyntaxFactory.TriviaList(CreateTriviaListFromTo(triviaList, index + 1, triviaList.Count - 1)))
+                    Return (SyntaxFactory.TriviaList(CreateTriviaListFromTo(triviaList, 0, index)), SyntaxFactory.TriviaList(CreateTriviaListFromTo(triviaList, index + 1, triviaList.Count - 1)))
                 End If
 
-                Return ValueTuple.Create(Of SyntaxTriviaList, SyntaxTriviaList)(GetSyntaxTriviaList(GetTextSpan(pair.Key), pair.Value, _cancellationToken), SyntaxTriviaList.Empty)
+                ' Grab the text change we're making and split it into the trailing trivia for the
+                ' previous token and the leading trivia for the next token.  The trivia may contain
+                ' multiple newlines, so we need to first grab the trailing portion (up through the
+                ' first newline), then use the remainder as the leading portion.
+                Dim text = pair.Value.GetTextChanges(GetTextSpan(pair.Key)).Single().NewText
+                Dim trailing = SyntaxFactory.ParseTrailingTrivia(text)
+                Dim leading = SyntaxFactory.ParseLeadingTrivia(text.Substring(trailing.FullSpan.Length))
+
+                Return (trailing, leading)
```

Comment:
this logic is different.  previously, the formatter assumed that the text it would get would only ever have at most 1 newline in it.  However,  system allows for rules to state they want multiple newlines.  This code was problematic as it would make nodes that then ended in `\r\n\r\n`.    This isn't how the actual trees are parsed, meaning we would produce trees differently than the parser.  

I augmented the testing system to actually validate that the trees we produce are the same as the trees we parse to catch this, and i discovered this. 

Tagging @heejaechang

Reply:
for vb, it is almost impossible to reconstruct how parser would created tree since unlike c#, vb doesn't have clear rule on which trivia attached to which token when parsing. I used to have test that validate reconstructed tree is same as one from parsing, but end up, remove those test because it is not possible for VB.

Reply:
is SyntaxFactory.ParseTrailingTrivia not designed for that purpose?